commons/src/main/java/org/infinispan/commons/api/functional/Listeners.java

```diff
@@ -0,0 +1,221 @@
+      /**
+       * Add a remove event specific listener by passing in a
+       * {@link Consumer} to be called back each time an entry is
+       * removed, passing in the {@link ReadEntryView} of the removed entry.
+       *
+       * This method is shortcut for users who are only interested in
+       * remove events. If interested in multiple event types, calling
+       * {@link #add(ReadWriteListener)} is recommended instead.
+       *
+       * @param f operation to be called each time a new cache entry is removed,
+       *          with the old cached entry's {@link ReadEntryView} as parameter.
+       * @return an {@link AutoCloseable} instance that can be used to
+       *         unregister the listener
+       */
+      AutoCloseable onRemove(Consumer<ReadEntryView<K, V>> f);
+
+      /**
+       * Add a read-write listener, and return an {@link AutoCloseable}
+       * instance that can be used to remove the listener registration.
+       *
+       * @param l the read-write functional map event listener
+       * @return an {@link AutoCloseable} instance that can be used to
+       *         unregister the listener
+       */
+      AutoCloseable add(ReadWriteListener<K, V> l);
+
+      /**
+       * Read-write listener
+       */
+      interface ReadWriteListener<K, V> {
+         /**
+          * Entry created event callback that receives a {@link ReadEntryView}
+          * of the created entry.
+          *
+          * @param created created entry view
+          */
+         default void onCreate(ReadEntryView<K, V> created) {}
+
+         /**
+          * Entry modify/update event callback that receives {@link ReadEntryView}
+          * of the previous entry as first parameter, and the {@link ReadEntryView}
+          * of the new entry.
+          *
+          * @param before previous entry view
+          * @param after new entry view
+          */
+         default void onModify(ReadEntryView<K, V> before, ReadEntryView<K, V> after) {}
+
+         /**
+          * Entry removed event callback that receives a {@link ReadEntryView}
+          * of the removed entry.
+          *
+          * @param removed removed entry view
+          */
+         default void onRemove(ReadEntryView<K, V> removed) {}
+      }
+   }
+
+   /**
+    * Write-only listeners enables user to register listeners for events
+    * happening in the write-only functional map.
+    *
+    * DESIGN RATIONALES:
+    * <ul>
+    *    <li>Write-only listeners cannot distinguish between cache entry
+    *    created and cache entry modify/update events. In either case, all
+    *    they know is that a new entry has been added, but they can't know if
+    *    entry was present before or not, nor what the previous value.
+    *    </li>
+    *    <li>Write-only listeners can distinguish between entry removals
+    *    vs cache entry create/modify-update events because they can query
+    *    what the new entry's value via {@link ReadEntryView#find()}.
```

Comment:
I was just disagreeing with the documentation saying that the listener can distinguish between removals and other kinds of modifications - it's a bit misleading because there may be no modification at all.